File: app/src/main/java/com/paradoxo/amadeus/util/GerenciaMusica.java
package com.paradoxo.amadeus.util;

import android.content.Context;
import android.database.Cursor;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.provider.MediaStore;
import android.util.Log;

import com.paradoxo.amadeus.modelo.Musica;
import com.paradoxo.amadeus.modelo.Sentenca;

import org.greenrobot.eventbus.EventBus;

public class GerenciaMusica {
    private Context context;
    private MediaPlayer mediaPlayer;

    public static final String REPRODUZINDO = "Reproduzindo ";

    public GerenciaMusica() {
    }

    public GerenciaMusica(Context context) {
        this.context = context;
    }

    public Musica encontrarMusica(String nomeMusicaBusada) {
        final Musica musica = new Musica();
        final boolean aleatorio = nomeMusicaBusada == null;
        if (nomeMusicaBusada == null) nomeMusicaBusada = "";
        final String ordernarAleatorio = aleatorio ? "RANDOM()" : null;

        Cursor cursor = context.getContentResolver().query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, null, null, null, ordernarAleatorio);
        if (cursor != null) {
            while (!cursor.isClosed() && cursor.moveToNext()) {
                String caminho = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DATA));
                String nomeMusica = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.DISPLAY_NAME));

                if (!caminho.contains("WhatsApp") && !caminho.contains("RecForge") && !caminho.contains("hangouts")) {
                    if (aleatorio || nomeMusica.toLowerCase().contains(nomeMusicaBusada)) {
                        musica.setNome(nomeMusica.substring(0, nomeMusica.lastIndexOf('.')));
                        musica.setCaminho(caminho);
                        musica.setArtista(cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.ARTIST)));
                        musica.setAlbum(cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM)));
                        String musicaAtual = musica.getNome();
                        break;
                    }
                } else if (aleatorio) {
                    if (!cursor.isClosed()) cursor.close();
                    encontrarMusica(null);
                }
            }
            if (!cursor.isClosed()) cursor.close();
        }

        return musica;
    }

    public void configurarMediPlayer(final Musica musica) {
        pararMusicaSeEstiverTocando();
        mediaPlayer = new MediaPlayer();
        mediaPlayer.setOnCompletionListener(mediaPlayer -> Log.e("A música", "Acabou"));

        try {
            mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
            try {
                mediaPlayer.setDataSource(musica.getCaminho());
            } catch (Exception e) {
                Log.e("Erro load", "Load musica");
                e.printStackTrace();
            }

            mediaPlayer.prepareAsync();
            mediaPlayer.setOnPreparedListener(mp -> tocarMusica(musica));

        } catch (Exception e) {
            Log.e("A música", "Não pode ser reproduzida");
        }
    }

    private void tocarMusica(Musica musica) {
        if (mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
        } else {
            mediaPlayer.start();
            Log.e(REPRODUZINDO, musica.getNome());
            notificarOutput(REPRODUZINDO + musica.getNome());
        }
    }

    public void pararMusicaSeEstiverTocando() {
        if (musicaEstaTocando()) {
            mediaPlayer.stop();
        }
    }

    public boolean musicaEstaTocando() {
        return mediaPlayer != null && mediaPlayer.isPlaying();
    }

    private static void notificarOutput(String mensagem) {
        Sentenca sentenca = new Sentenca(mensagem);
        EventBus.getDefault().post(sentenca);
    }
}

File: app/src/main/java/com/paradoxo/amadeus/util/GerenciaMusica.kt
package com.paradoxo.amadeus.util

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.provider.MediaStore
import android.util.Log
import com.paradoxo.amadeus.modelo.Musica
import com.paradoxo.amadeus.modelo.Sentenca
import org.greenrobot.eventbus.EventBus

class GerenciaMusica(private val context: Context? = null) {
    private var mediaPlayer: MediaPlayer? = null

    companion object {
        const val REPRODUZINDO = "Reproduzindo "
    }

    fun encontrarMusica(nomeMusicaBuscada: String?): Musica {
        val musica = Musica()
        val aleatorio = nomeMusicaBuscada == null
        val nomeBusca = nomeMusicaBuscada ?: ""
        val ordenarAleatorio = if (aleatorio) "RANDOM()" else null

        val cursor = context?.contentResolver?.query(
            MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, null, null, null, ordenarAleatorio
        )
        cursor?.use {
            while (!it.isClosed && it.moveToNext()) {
                val caminho = it.getString(it.getColumnIndexOrThrow(MediaStore.Audio.Media.DATA))
                val nomeMusica = it.getString(it.getColumnIndexOrThrow(MediaStore.Audio.Media.DISPLAY_NAME))

                if (!caminho.contains("WhatsApp") && !caminho.contains("RecForge") && !caminho.contains("hangouts")) {
                    if (aleatorio || nomeMusica.lowercase().contains(nomeBusca)) {
                        musica.nome = nomeMusica.substring(0, nomeMusica.lastIndexOf('.'))
                        musica.caminho = caminho
                        musica.artista = it.getString(it.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST))
                        musica.album = it.getString(it.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM))
                        break
                    }
                } else if (aleatorio) {
                    if (!it.isClosed) it.close()
                    encontrarMusica(null)
                }
            }
        }
        return musica
    }

    fun configurarMediPlayer(musica: Musica) {
        pararMusicaSeEstiverTocando()
        mediaPlayer = MediaPlayer().apply {
            setOnCompletionListener { Log.e("A música", "Acabou") }
            try {
                @Suppress("DEPRECATION")
                setAudioStreamType(AudioManager.STREAM_MUSIC)
                try {
                    setDataSource(musica.caminho)
                } catch (e: Exception) {
                    Log.e("Erro load", "Load musica")
                    e.printStackTrace()
                }
                prepareAsync()
                setOnPreparedListener { tocarMusica(musica) }
            } catch (e: Exception) {
                Log.e("A música", "Não pode ser reproduzida")
            }
        }
    }

    private fun tocarMusica(musica: Musica) {
        mediaPlayer?.let {
            if (it.isPlaying) it.stop()
            else {
                it.start()
                Log.e(REPRODUZINDO, musica.nome)
                notificarOutput(REPRODUZINDO + musica.nome)
            }
        }
    }

    fun pararMusicaSeEstiverTocando() {
        if (musicaEstaTocando()) mediaPlayer?.stop()
    }

    fun musicaEstaTocando(): Boolean = mediaPlayer?.isPlaying == true

    private fun notificarOutput(mensagem: String) {
        EventBus.getDefault().post(Sentenca(mensagem))
    }
}
